app/uk/gov/hmrc/fileupload/ApplicationModule.scala

```diff
@@ -0,0 +1,193 @@
+
+import java.util.UUID
+
+import akka.actor.ActorRef
+import com.kenshoo.play.metrics.MetricsImpl
+import javax.inject.{Inject, Singleton}
+import play.api._
+import play.api.libs.ws.ahc.AhcWSComponents
+import play.modules.reactivemongo.ReactiveMongoComponent
+import reactivemongo.api.commands.WriteConcern
+import uk.gov.hmrc.fileupload.controllers.RetrieveFile
+import uk.gov.hmrc.fileupload.file.zip.Zippy
+import uk.gov.hmrc.fileupload.infrastructure._
+import uk.gov.hmrc.fileupload.read.envelope.{WithValidEnvelope, Service => EnvelopeService, _}
+import uk.gov.hmrc.fileupload.read.notifier.{NotifierActor, NotifierRepository}
+import uk.gov.hmrc.fileupload.read.routing.{RoutingActor, RoutingConfig, RoutingRepository}
+import uk.gov.hmrc.fileupload.read.stats.{Stats, StatsActor, StatsLogWriter, StatsLogger, StatsLoggingConfiguration, StatsLoggingScheduler, Repository => StatsRepository}
+import uk.gov.hmrc.fileupload.write.envelope._
+import uk.gov.hmrc.fileupload.write.infrastructure.UnitOfWorkSerializer.{UnitOfWorkReader, UnitOfWorkWriter}
+import uk.gov.hmrc.fileupload.write.infrastructure.{Aggregate, MongoEventStore, StreamId}
+import uk.gov.hmrc.lock.LockRepository
+import uk.gov.hmrc.play.audit.http.connector.AuditConnector
+import uk.gov.hmrc.play.bootstrap.config.ServicesConfig
+
+@Singleton
+class ApplicationModule @Inject()(
+  servicesConfig: ServicesConfig,
+  reactiveMongoComponent: ReactiveMongoComponent,
+  auditConnector: AuditConnector,
+  metrics: MetricsImpl,
+  val applicationLifecycle: play.api.inject.ApplicationLifecycle,
+  val configuration: play.api.Configuration,
+  val environment: play.api.Environment,
+  actorSystem: akka.actor.ActorSystem
+)(implicit
+  val executionContext: scala.concurrent.ExecutionContext,
+  val materializer: akka.stream.Materializer
+) extends AhcWSComponents {
+
+  private val logger = Logger(getClass)
+
+  lazy val db = reactiveMongoComponent.mongoConnector.db
+
+  implicit val reader = new UnitOfWorkReader(EventSerializer.toEventData)
+  implicit val writer = new UnitOfWorkWriter(EventSerializer.fromEventData)
+
+  val envelopeConstraintsConfigure: EnvelopeConstraintsConfiguration =
+    EnvelopeConstraintsConfiguration.getEnvelopeConstraintsConfiguration(configuration) match {
+      case Right(envelopeConstraints) => envelopeConstraints
+      case Left(failureReason) => throw new IllegalArgumentException(s"${failureReason.message}")
+    }
+
+  val envelopeHandler = new EnvelopeHandler(envelopeConstraintsConfigure)
+
+  val subscribe: (ActorRef, Class[_]) => Boolean = actorSystem.eventStream.subscribe
+  val publish: (AnyRef) => Unit = actorSystem.eventStream.publish
+  val withBasicAuth: BasicAuth = BasicAuth(basicAuthConfiguration(configuration))
+
+  val eventStore = if (configuration.get[Boolean]("mongodb.replicaSetInUse")) {
+    new MongoEventStore(db, metrics.defaultRegistry, writeConcern = WriteConcern.ReplicaAcknowledged(n = 2, timeout = 5000, journaled = true))
```

Comment:
was wondering, if it needs 2 replicas to ack the write, what happens if one of the replicas is down/recycling?